clang/lib/CodeGen/CGCall.cpp

```diff
@@ -4148,6 +4151,30 @@ static void emitWriteback(CodeGenFunction &CGF,
   assert(!isProvablyNull(srcAddr.getBasePointer()) &&
          "shouldn't have writeback for provably null argument");
 
+  if (CGF.getLangOpts().HLSL) {
+    if (writeback.CastExpr) {
+      RValue TmpVal = CGF.EmitAnyExprToTemp(writeback.CastExpr);
+      if (TmpVal.isScalar())
+        CGF.EmitStoreThroughLValue(TmpVal, srcLV);
+      else
+        CGF.EmitAggregateStore(srcLV.getPointer(CGF),
```

Comment:
I think there's a codegen bug (and missing test coverage related to this. I'm debugging right now and will fix and post additional tests once I have it resolved.

Reply:
This should be resolved and covered in the updated tests.